File: solutions/cpp/0368.cpp
class Solution {
 public:
  vector<int> largestDivisibleSubset(vector<int>& nums) {
    const int n = nums.size();

    vector<int> ans;
    vector<int> lengthEndsAt(n, 1);
    vector<int> prevIndex(n, -1);
    int maxLength = 0;
    int index = -1;

    sort(begin(nums), end(nums));

    for (int i = 0; i < n; ++i) {
      for (int j = i - 1; j >= 0; --j)
        if (nums[i] % nums[j] == 0 && lengthEndsAt[i] < lengthEndsAt[j] + 1) {
          lengthEndsAt[i] = lengthEndsAt[j] + 1;
          prevIndex[i] = j;
        }
      if (lengthEndsAt[i] > maxLength) {
        maxLength = lengthEndsAt[i];
        index = i;
      }
    }

    while (index != -1) {
      ans.push_back(nums[index]);
      index = prevIndex[index];
    }

    return ans;
  }
};

File: solutions/java/0368.java
class Solution {
  public List<Integer> largestDivisibleSubset(int[] nums) {
    final int n = nums.length;

    List<Integer> ans = new ArrayList<>();
    int[] count = new int[n];
    int[] prevIndex = new int[n];
    int maxCount = 0;
    int index = -1;

    Arrays.fill(count, 1);
    Arrays.fill(prevIndex, -1);
    Arrays.sort(nums);

    for (int i = 0; i < n; ++i) {
      for (int j = i - 1; j >= 0; --j)
        if (nums[i] % nums[j] == 0 && count[i] < count[j] + 1) {
          count[i] = count[j] + 1;
          prevIndex[i] = j;
        }
      if (count[i] > maxCount) {
        maxCount = count[i];
        index = i;
      }
    }

    while (index != -1) {
      ans.add(nums[index]);
      index = prevIndex[index];
    }

    return ans;
  }
}